keras_cv/losses/giou_loss.py

```diff
@@ -0,0 +1,151 @@
+                f"`len(boxes2.shape)`={boxes2_rank}.  Expected either `len(boxes1.shape)`=2 AND "
+                "`len(boxes2.shape)`=2, or `len(boxes1.shape)`=3 AND `len(boxes2.shape)`=3."
+            )
+
+        if self.bounding_box_format.startswith("rel"):
+            target = "rel_yxyx"
+        else:
+            target = "yxyx"
+
+        boxes1 = bounding_box.convert_format(
+            boxes1, source=self.bounding_box_format, target=target
+        )
+
+        boxes2 = bounding_box.convert_format(
+            boxes2, source=self.bounding_box_format, target=target
+        )
+
+        def compute_giou_for_batch(boxes):
+            boxes1, boxes2 = boxes
+            zero = tf.convert_to_tensor(0.0, boxes1.dtype)
+            boxes1_ymin, boxes1_xmin, boxes1_ymax, boxes1_xmax = tf.unstack(
+                boxes1[..., :4], 4, axis=-1
+            )
+            boxes2_ymin, boxes2_xmin, boxes2_ymax, boxes2_xmax = tf.unstack(
+                boxes2[..., :4], 4, axis=-1
+            )
+            boxes1_width = tf.maximum(zero, boxes1_xmax - boxes1_xmin)
+            boxes1_height = tf.maximum(zero, boxes1_ymax - boxes1_ymin)
+            boxes2_width = tf.maximum(zero, boxes2_xmax - boxes2_xmin)
+            boxes2_height = tf.maximum(zero, boxes2_ymax - boxes2_ymin)
+            boxes1_area = boxes1_width * boxes1_height
+            boxes2_area = boxes2_width * boxes2_height
+            intersect_ymin = tf.maximum(boxes1_ymin, boxes2_ymin)
+            intersect_xmin = tf.maximum(boxes1_xmin, boxes2_xmin)
+            intersect_ymax = tf.minimum(boxes1_ymax, boxes2_ymax)
+            intersect_xmax = tf.minimum(boxes1_xmax, boxes2_xmax)
+            intersect_width = tf.maximum(zero, intersect_xmax - intersect_xmin)
+            intersect_height = tf.maximum(zero, intersect_ymax - intersect_ymin)
+            intersect_area = intersect_width * intersect_height
+
+            union_area = boxes1_area + boxes2_area - intersect_area
+            iou = tf.math.divide_no_nan(intersect_area, union_area)
+
+            # giou calculation
+            enclose_ymin = tf.minimum(boxes1_ymin, boxes2_ymin)
+            enclose_xmin = tf.minimum(boxes1_xmin, boxes2_xmin)
+            enclose_ymax = tf.maximum(boxes1_ymax, boxes2_ymax)
+            enclose_xmax = tf.maximum(boxes1_xmax, boxes2_xmax)
+            enclose_width = tf.maximum(zero, enclose_xmax - enclose_xmin)
+            enclose_height = tf.maximum(zero, enclose_ymax - enclose_ymin)
+            enclose_area = enclose_width * enclose_height
+            return iou - tf.math.divide_no_nan(
+                (enclose_area - union_area), enclose_area
+            )
+
+        if boxes1_rank == 2:
+            return compute_giou_for_batch((boxes1, boxes2))
+        else:
+            return tf.vectorized_map(compute_giou_for_batch, elems=(boxes1, boxes2))
+
+    def call(self, y_true, y_pred):
```

Comment:
support `sample_weight`?

Reply:
On it

Reply:
Sorry for the delay on this. 

@tanzhenyu Is there a specific reason we need to add this? It's inbuilt in Keras and Addons implementations uses the `LossFunctionWrapper`. Is this worth the complex code we will need to add to support it?

Reply:
But this loss is not inhering from `LossFunctionWrapper`, correct?

Reply:
Yeah, its not